### ISSUE TITLE
How to enable Hydra delegate?

### ISSUE THREAD

**Issue by BrianHanke** [dreamworksanimation/openmoonray#48]
Doing good over here, managed to compile MoonRay on a CentoOS 7 VM, copied the build over to my main RHEL 8 VM and everything is actually working! (After messing with some libraries.)

Over in Houdini I've got all the MoonRay nodes and parameters showing up in Solaris, but I can't figure out how to enable the Hydra delegate. I tried a bunch of things in houdini.env, but no go. Any tips?

![yay](https://user-images.githubusercontent.com/59420805/226147531-23f09804-e0b1-48e7-84b5-75d46871aeb5.png)
![progress](https://user-images.githubusercontent.com/59420805/226147572-1a9aab3f-c2e8-4238-85e0-d309ad5ebf20.png)

**Comment by BrianHanke** [comment]
Yeah I was wondering if it needs to be built against Houdini. Also, I have the latest production build installed on my VM, maybe DreamWorks uses an older one?

Would be nice if the build docs included this since I assume 99.9% of us want to use MoonRay in Houdini.

**Comment by nospitters** [comment]
yes, hoping there will be a more detailed instructions for houdini setup/install...
and 99% of the 99.9% wants it in windows 🤪

**Comment by rgwilson68** [comment]
What I have seen is the Houdini shows MoonRay in the list of renderers if it finds the plugin : if it disappears after you try to switch to that means loading of hdMoonray.so failed, and you can see the error details by clicking on the red warning triangle.

Problems finding the plugin are probably to do with the setting of PXR_PLUGINPATH_NAME : you should have something like this:
> echo $PXR_PLUGINPATH_NAME
/installs/openmoonray/scripts/../plugin/usd:
> ls /installs/openmoonray/scripts/../plugin/usd
hdMoonray  hdMoonrayAdapters  moonrayShaderDiscovery  moonrayShaderParser  plugInfo.json

For libraries that are used by both Houdini and MoonRay, I think you will need to build against the versions that Houdini uses : otherwise hdMoonray.so will find the Houdini versions on LD_LIBRARY_PATH at runtime and may fail to load. The viewport messages in Houdini should show which library was incompatible.

The version of Usd that Houdini uses is available here: https://github.com/sideeffects/USD. It should be possible to build against the actual headers and libraries shipped with Houdini, but you would need a pxrConfig.cmake file to configure Usd properly, and I don't think that is included.

**Comment by nospitters** [comment]
here i see that you also include the ../scripts path to the pxr_pluginpath_name. is that relevant?

houdini doesnt even show up moonray and it doesnt spit out any errors...

from what i can see is that moonray is build with usd 22.05...houdini 19.5 is using usd 22.05a...have the usd versions to match exactly?
https://www.sidefx.com/docs/houdini/news/19_5/platforms.html

**Comment by BrianHanke** [comment]
That latest production build gives no details on the error, but I tried installing a slightly older version of 19.5 and that one gives a clue:

Hydra: failed to load plugin 'hdMoonray': libusd_usdImaging.so: cannot open shared object file: No such file or directory in '/install/openmoonray/plugin/hdMoonray.so'

I can confirm that libusd_usdImaging.so is indeed not in the plugin folder and is not even anywhere in the openmoonray folders. Should it be?

**Comment by nospitters** [comment]
@rgwilson68 thx for debugging...have to try it out tomorrow...

**Comment by BrianHanke** [comment]
Thanks for the great info @rgwilson68. Will keep running tests over here.

**Comment by BrianHanke** [comment]
Well, I sorted through a bunch of viewport errors with all the USD .so files that were missing from the MoonRay build and needed to be symlinked from the "pxr" versions that come with Houdini. That was all working fine, but then I hit this one:

/lib64/libk5crypto.so.3: undefined symbol: EVP_KDF_ctrl, version OPENSSL_1_1_1b

My VM has openssl1.1.1k and it seems like MoonRay is looking for 1.1.1b? If so, that's the end of the line for me since force installing an older version removes like half my installed packages. Would have to start from scratch with a whole new Linux build.

**Comment by nospitters** [comment]
tried all the recommendations above, also copied the libusd_usdImaging.so (which lives in /usr/local/lib/ in my case) side to hdMoonray.so...but all taht desnt help.
set the TF_DEBUG="PLUG_*", also installed houdini 19.5.303 to see if i get any errors, but houdini does not spits anything to the console, and doesnt show the renderer in solaris...
so i give up here and waiting for a maybe function prebuild-release....

thx to everybody...

**Comment by panda1100** [comment]
@BrianHanke Do you have any notes for setting up Houdini + HdMoonRay, by any chance? I don't mind if it's not complete procedures and/or including trial & errors. I have HPC eng. background but super new to rendering and doesn't have Houdini yet 😅 I built MonnRay on Rocky Linux 9 and tested multi node rendering on my local environment so far. Now I'm very curious about real use-case like Houdini + MoonRay.

**Comment by BrianHanke** [comment]
@panda1100 Yes! I released a build for Houdini 19.5.534 on CentOS 7 here: https://github.com/BrianHanke/openmoonray-bhgc/releases/tag/v2023.4

I also wrote a few guides and articles here: https://github.com/BrianHanke/openmoonray-bhgc/discussions

**Comment by panda1100** [comment]
Thank you @BrianHanke! It will definitely help :)
How did you find this file? https://docs.openmoonray.org/assets/build_extras/moonray_houdini_pxr_config.tar

**Comment by BrianHanke** [comment]
> How did you find this file?

That was shared elsewhere here by Rob Wilson. Would never have figured it out without him!

**Comment by panda1100** [comment]
@BrianHanke I found 2 versions of Houdini installer, what is the difference between the two?  python3 version of installer looks smaller than non python3 one..

[houdini-19.5.569-linux_x86_64_gcc9.3.tar.gz ](https://www.sidefx.com/download/eula/accept/?next=/download/download-houdini/100769/) 2.3 GB
[houdini-py37-19.5.569-linux_x86_64_gcc9.3.tar.gz ](https://www.sidefx.com/download/eula/accept/?next=/download/download-houdini/100788/) 2.2 GB

**Comment by BrianHanke** [comment]
@panda1100 My build was based on [houdini-19.5.534-linux_x86_64_gcc9.3.tar.gz](https://www.sidefx.com/download/download-houdini/99589/). Default Houdini is Python 3.9, but they also offer a 3.7 option.

**Comment by panda1100** [comment]
@BrianHanke Thank you!!!

**Comment by mpjohn** [comment]
Hi There, Long time Maya user - brand new to Houdini.
Are there instructions anywhere for how to import the Dwa shaders into Houdini. Is this through the Hydra plugin?
Thanks
M

**Comment by golfingmateo** [comment]
If you have everything included correctly then you should find there are Moonray* Shader HDA's available in the matnet.

You can assign shaders with the LOP material library

**Comment by golfingmateo** [comment]
if the shaders are not showing up make sure you have moved the HDA's into a directory sourced by HOUDINI.

**Comment by jlanz** [comment]
Take a look at this:
https://docs.openmoonray.org/user-reference/tools/hydra/hdmoonray-houdini/

**Comment by mpjohn** [comment]
Thanks jlanz - that's exactly what I was looking for. After setting up Moonray I thought Houdini setup ( licensing etc ) would be easy...boy was I wrong!

**Comment by cedricp** [comment]
I finally got moonray compiled and working in Houdini (19.5) with the Hydra delegate. The only problem I have is that there are no editable parameters for materials, it looks like the .ds files are not found/used (parameter window empty). I tried to copy files to the Houdini user folder and setting HOUDINI_PATH env var correctly, but both aren't working... RenderSettings node and HdMoonrayRendererPlugin_Viewport are found, though. I also tried to enable houini debug but nothing relevant displayed.
